File: python/0076_minimum-window-substring.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Author        : Rock Wayne 
# @Created       : 2020-05-06 08:00:00
# @Last Modified : 2020-05-06 08:00:00
# @Mail          : lostlorder@gmail.com
# @Version       : alpha-1.0

"""
# 给你一个字符串 S、一个字符串 T，请在字符串 S 里面找出：包含 T 所有字母的最小子串。 
# 
#  示例： 
# 
#  输入: S = "ADOBECODEBANC", T = "ABC"
# 输出: "BANC" 
# 
#  说明： 
# 
#  
#  如果 S 中不存这样的子串，则返回空字符串 ""。 
#  如果 S 中存在这样的子串，我们保证它是唯一的答案。 
#  
#  Related Topics 哈希表 双指针 字符串 Sliding Window

"""
import collections

import pytest


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    """
    滑动窗口
    https://leetcode-cn.com/problems/minimum-window-substring/solution/hua-dong-chuang-kou-suan-fa-tong-yong-si-xiang-by-/
    """

    def minWindow(self, s: str, t: str) -> str:
        """Good"""
        if not t or not s:
            return ""
        # Dictionary which keeps a count of all the unique characters in t.
        dict_t = collections.Counter(t)
        # Number of unique characters in t, which need to be present in the desired window.
        required = len(dict_t)
        # left and right pointer
        l, r = 0, 0
        # formed is used to keep track of how many unique characters in t are present in the current window in its desired frequency.
        # e.g. if t is "AABC" then the window must have two A's, one B and one C. Thus formed would be = 3 when all these conditions are met.
        formed = 0
        # Dictionary which keeps a count of all the unique characters in the current window.
        window_counts = {}
        # ans tuple of the form (window length, left, right)
        ans = float("inf"), None, None

        while r < len(s):
            # Add one character from the right to the window
            character = s[r]
            window_counts[character] = window_counts.get(character, 0) + 1

            # If the frequency of the current character added equals to the desired count in t then increment the formed count by 1.
            if character in dict_t and window_counts[character] == dict_t[character]:
                formed += 1
            # Try and contract the window till the point where it ceases to be 'desirable'.
            while l <= r and formed == required:
                character = s[l]

                # Save the smallest window until now.
                if r - l + 1 < ans[0]:
                    ans = (r - l + 1, l, r)
                # The character at the position pointed by the `left` pointer is no longer a part of the window.
                window_counts[character] -= 1
                if character in dict_t and window_counts[character] < dict_t[character]:
                    formed -= 1
                # Move the left pointer ahead, this would help to look for a new window.
                l += 1
            # Keep expanding the window once we are done contracting.
            r += 1
        return "" if ans[0] == float("inf") else s[ans[1]: ans[2] + 1]


# leetcode submit region end(Prohibit modification and deletion)

class Solution1:

    def minWindow(self, s: str, t: str) -> str:
        score = 0
        wanted = collections.Counter(t)
        start, end = len(s), 3 * len(s)
        d = collections.defaultdict(int)
        deq = collections.deque([])
        for idx, char in enumerate(s):
            if char in wanted:
                deq.append(idx)
                d[char] += 1
                if d[char] <= wanted[char]:
                    score += 1
                while deq and d[s[deq[0]]] > wanted[s[deq[0]]]:
                    d[s[deq.popleft()]] -= 1
                if score == len(t) and deq[-1] - deq[0] < end - start:
                    start, end = deq[0], deq[-1]
        return s[start:end + 1]


class Solution2:

    def minWindow(self, s: str, t: str) -> str:
        """
        https://leetcode-cn.com/problems/minimum-window-substring/solution/python-jian-ji-hua-dong-chuang-kou-xiang-jie-by-am/
        """
        counter = collections.Counter(t)
        ans = ""
        cnt = 0  # 当前我满足了 t 中的字母的种数
        l = 0
        for r, r_char in enumerate(s):
            if r_char not in counter:
                continue
            counter[r_char] -= 1
            if counter[r_char] == 0:
                cnt += 1
            while s[l] not in counter or counter[s[l]] < 0:  # 看看当前 l 处的字母是否必要，没必要 l 就加以
                if s[l] in counter:
                    counter[s[l]] += 1
                l += 1
            if cnt == len(counter):
                if not ans or len(ans) > r - l + 1:
                    ans = s[l: r + 1]
        return ans


@pytest.mark.parametrize("kwargs,expected", [
    (dict(s="ADOBECODEBANC", t="ABC"), "BANC"),
])
def test_solutions(kwargs, expected):
    assert Solution().minWindow(**kwargs) == expected
    assert Solution1().minWindow(**kwargs) == expected
    assert Solution2().minWindow(**kwargs) == expected


if __name__ == '__main__':
    pytest.main(["-q", "--color=yes", "--capture=no", __file__])

File: C/src/0076_minimum-window-substring.c
//
/* Created by rock on 5/2/20.
*/
#include "leetcode_functions.h"

//https://leetcode-cn.com/problems/minimum-window-substring/solution/shuang-100-cyu-yan-hua-dong-chuang-kou-by-xiao-xi-/


char *minWindow(char *s, char *t) {

    if (s == NULL || t == NULL) {
        return NULL;
    }
    int lens = (int) strlen(s);
    int lent = (int) strlen(t);
    char *out = (char *) malloc(lens + 1);
    if (out == NULL) {
        return NULL;
    }
    int outLen = 0;
    memset(out, 0, lens + 1);
    if (lens < lent || lent == 0) {
        return out;
    }
    int map_s[128] = {0};
    int map_t[128] = {0};
    int cnt = 0;
    for (int i = 0; i < lent; ++i) {
        map_t[t[i]]++;
        map_s[s[i]]++;
    }
    for (int i = 0; i < 128; ++i) {
        if (map_t[i] > 0) {
            /* 临界点，只要 <= 短串 就是有效计数, 再大就是多余的了 */
            cnt += min(map_t[i], map_s[i]);
        }

    }

    /* 2. 当计数等于短串长度就是满足条件了 */
    if (cnt == lent) {
        outLen = lent;
        memcpy(out, s, outLen);
        return out;
    }
    /* 3. 滑动窗口 */
    int l = 0;
    int r = lent;

    for (; r < lens; r++) {
        if (map_t[s[r]] == 0) {
            continue;
        }
        map_s[s[r]]++;
        if (map_s[s[r]] <= map_t[s[r]]) {
            /* 临界点，只要 <= 短串 就是有效计数, 再大就是多余的了 */

            cnt++;
        }
        if (cnt == lent) {
            /* 从左侧开始校验 找到临界点（map[s[l]] == mapt[s[l]]）删除多余数据 */
            while (map_t[s[l]] == 0 || map_s[s[l]] > map_t[s[l]]) {
                map_s[s[l++]]--;
            }
            int cpLen = r - l + 1;
            if (outLen == 0 || cpLen < outLen) {
                memset(out, 0, lens + 1);
                memcpy(out, s + l, cpLen);
                outLen = cpLen;
            }
            /* 临界点 */

            cnt--;
            map_s[s[l++]]--;

        }


    }
    return out;


}